node/src/components/consensus/era_supervisor.rs

```diff
@@ -443,32 +445,72 @@ where
-                block_header.timestamp(),
-                block_header.height() + 1,
-                *block_header.global_state_hash(),
+            // if the block is a switch block, we have to get the validators for the new era and
+            // create it, before we can say we handled the block
+            let request = GetEraValidatorsRequest::new(
+                (*block_header.global_state_hash()).into(),
+                block_header.era_id().successor().0,
+                ProtocolVersion::V1_0_0,
+            );
+            effects.extend(self.effect_builder.get_validators(request).event(|result| {
+                Event::GetValidatorsResponse {
+                    switch_block_header: block_header,
+                    get_validators_result: result,
+                }
+            }));
+        } else {
+            // if it's not a switch block, we can already declare it handled
+            effects.extend(
+                self.effect_builder
+                    .announce_block_handled(block_header)
+                    .ignore(),
             );
-            effects.extend(self.handle_consensus_results(new_era_id, results));
         }
-        effects.extend(
-            self.effect_builder
-                .announce_block_handled(block_header)
-                .ignore(),
-        );
         effects
     }
 
+    pub(super) fn handle_validators_response(
+        &mut self,
+        block_header: BlockHeader,
+        validators: Option<ValidatorWeights>,
+    ) -> Effects<Event<I>> {
+        match validators {
+            Some(validator_weights) => {
+                let validator_stakes = validator_weights
+                    .into_iter()
+                    .map(|(key, stake)| {
+                        (
+                            key.try_into().ok().expect("keys should convert correctly"),
+                            Motes::new(stake),
+                        )
+                    })
+                    .collect();
+                self.era_supervisor
+                    .current_era_mut()
+                    .consensus
+                    .deactivate_validator();
+                let new_era_id = block_header.era_id().successor();
+                info!(?new_era_id, "Era created");
+                let results = self.era_supervisor.new_era(
+                    new_era_id,
+                    Timestamp::now(), // TODO: This should be passed in.
+                    validator_stakes,
+                    block_header.timestamp(),
+                    block_header.height() + 1,
+                    *block_header.global_state_hash(),
+                );
+                let mut effects = self.handle_consensus_results(new_era_id, results);
+                effects.extend(
+                    self.effect_builder
+                        .announce_block_handled(block_header)
+                        .ignore(),
+                );
+                effects
+            }
+            None => todo!(),
```

Comment:
I guess this will be a panic, or at least an error. We could return early, instead of having this `None =>` arm after the long main part of the method.